internal/api/v1/organizations.go

```diff
@@ -161,3 +155,53 @@ func (oc OrganizationsController) Delete(w http.ResponseWriter, r *http.Request)
 
 	return nil
 }
+
+func deleteApps(cluster *kubernetes.Cluster, gitea *gitea.Client, org string) error {
+	apps, err := application.List(cluster, org)
+	if err != nil {
+		return err
+	}
+
+	const maxConcurrent = 100
+	errChan := make(chan error, maxConcurrent)
+	success := make(chan struct{})
+	stop := make(chan struct{})
+
+	go func() {
+		var wg sync.WaitGroup
+		// create a buffer, so there is no more than 'maxConcurrent' goroutines running at the same time
+		buffer := make(chan struct{}, maxConcurrent)
+	loop:
+		for _, app := range apps {
+			buffer <- struct{}{}
+
+			go func(app application.Application) {
+				defer wg.Done()
+				wg.Add(1)
+				err = application.Delete(cluster, gitea, org, app)
+				if err != nil {
+					errChan <- err
+				}
+				<-buffer
+			}(app)
+
+			select {
+			case <-stop:
+				break loop
+			default:
+			}
+		}
+		wg.Wait()
+		success <- struct{}{}
+	}()
+
+	//wait until all apps are deleted or we get an error
+	select {
+	case err := <-errChan:
+		stop <- struct{}{}
```

Comment:
This will stop the loop so no more applications will be added for deletion (in the buffer) but we may have up to 99 applications currently being deleted. We should better wait for those go routines to finish. We can achieve that by moving the declaration of the wg variable (https://github.com/epinio/epinio/pull/460/files#diff-27a9ad008b6be81b374b2666592a4f2a279860da73cceaccf0268140cbf94686R171) ouside the main go routine (so that it's in scope here). Then we can simply defer the `wg.Wait()` call in the main thread (ideally right after the declaration of the variable).

Reply:
All gouroutines in the buffer will finished before any code after wg.Wait() is executed, so current deletions will finish